src/controller/java/BdxOTASender.cpp

```diff
@@ -0,0 +1,411 @@
+
+    if (mExchangeCtx != nullptr)
+    {
+        mExchangeCtx->Close();
+        mExchangeCtx = nullptr;
+    }
+
+    mInitialized = false;
+}
+
+CHIP_ERROR BdxOTASender::OnMessageToSend(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(mExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mOtaDelegate != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    Messaging::SendFlags sendFlags;
+
+    // All messages sent from the Sender expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    if (!event.msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse);
+    }
+
+    auto & msgTypeData = event.msgTypeData;
+    // If there's an error sending the message, close the exchange and call ResetState.
+    // TODO: If we can remove the !mInitialized check in ResetState(), just calling ResetState() will suffice here.
+    CHIP_ERROR err =
+        mExchangeCtx->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(event.MsgData), sendFlags);
+    if (err != CHIP_NO_ERROR)
+    {
+        mExchangeCtx->Close();
+        mExchangeCtx = nullptr;
+        ResetState();
+    }
+    else if (event.msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        // If the send was successful for a status report, since we are not expecting a response the exchange context is
+        // already closed. We need to null out the reference to avoid having a dangling pointer.
+        mExchangeCtx = nullptr;
+        ResetState();
+    }
+    return err;
+}
+
+bdx::StatusCode BdxOTASender::GetBdxStatusCodeFromChipError(CHIP_ERROR err)
+{
+    if (err == CHIP_ERROR_INCORRECT_STATE)
+    {
+        return bdx::StatusCode::kUnexpectedMessage;
+    }
+    if (err == CHIP_ERROR_INVALID_ARGUMENT)
+    {
+        return bdx::StatusCode::kBadMessageContents;
+    }
+    return bdx::StatusCode::kUnknown;
+}
+
+CHIP_ERROR BdxOTASender::OnTransferSessionBegin(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+    // Once we receive the BDX init, cancel the BDX Init timeout and start the BDX session
+    if (mSystemLayer)
+    {
+        mSystemLayer->CancelTimer(HandleBdxInitReceivedTimeoutExpired, this);
+    }
+
+    VerifyOrReturnError(mFabricIndex != kUndefinedFabricIndex, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mNodeId != kUndefinedNodeId, CHIP_ERROR_INCORRECT_STATE);
+    uint16_t fdl = 0;
+
+    const uint8_t * fd = mTransfer.GetFileDesignator(fdl);
+    VerifyOrReturnError(fdl <= bdx::kMaxFileDesignatorLen, CHIP_ERROR_INVALID_ARGUMENT);
+    CharSpan fileDesignatorSpan(Uint8::to_const_char(fd), fdl);
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    UtfString fileDesignator(env, fileDesignatorSpan);
```

Comment:
UtfString class is problematic, when calling from native layer, UtfString is casting local to global, but cannot release local, which would cause local reference leaks.
The recommended pattern is as below. I have converted this for android callback cpp file, need to spread this pattern to all jni file, at this moment, UTFSting and ByteArray..in JniWrapper should not be used now..
Pattern:
JniLocalReferenceManager manager(env); ==> manage the lifetime for all local refs in current function.
....
jstring fileDesignator = env->NewStringUTF(xx) :

Reply:
I add "chip::JniLocalReferenceManager manager(env)" code in this function